File: backend/internal/domain/story.go
package domain

import "time"

// Story is a full stories row.
type Story struct {
	ID, AuthorID, MediaID int64
	// Seq — номер истории ВНУТРИ автора: то, чем она адресуется снаружи.
	// Внутренний ключ строки (`ID`) наружу не выходит — приём и причина те же,
	// что у сообщения (`messages.seq` против `messages.id`).
	Seq                  int64
	Caption, Privacy     string
	CreatedAt, ExpiresAt time.Time
	// Pinned — история закреплена в профиле (показывается всегда, в т.ч. истёкшая).
	// Edited — история редактировалась после публикации (флаг edited в tweb).
	Pinned, Edited bool
	// MediaAreas — интерактивные области поверх истории (`Vector<MediaArea>`).
	MediaAreas MediaAreas
	// FwdFrom — ссылка на исходную историю при репосте (tweb fwd_from); nil у оригинала.
	FwdFrom *StoryFwd
}

// StoryFwd — ссылка репоста на исходную историю (tweb fwd_from): автор и НОМЕР
// оригинала внутри него. Глобального ключа здесь нет: на проводе
// `storyFwdHeader{from, story_id}` адресует пиром плюс номером.
type StoryFwd struct {
	AuthorID int64
	StoryID  int64
}

// StoryOrigin — минимальные данные исходной истории для репоста (переиспользуем
// media) и share в чаты (атрибуция по имени автора). domain.ErrNotFound если нет.
type StoryOrigin struct {
	AuthorID   int64
	AuthorName string
	MediaID    int64
}

// StoryRecord — ПЛОСКАЯ строка выборки историй: то, что отдаёт SQL, вместе с
// пер-зрительским состоянием (просмотрено, моя реакция).
//
// Суффикс Record — тот же приём и та же причина, что у DialogRecord/UserRecord/
// ChatRecord: `StoryItem` это имя ОБЪЕДИНЕНИЯ схемы (mtstory.go), и плоская
// строка выборки объединением не является. Разбор полей (что куда уезжает) —
// docs/readiness/tl-stories-analysis.md, исполняется шагом B.
type StoryRecord struct {
	ID, MediaID int64
	// Seq — номер внутри автора; наружу едет он, а не ключ строки.
	Seq     int64
	Caption string
	// Privacy отдаётся как есть, чтобы фронт показал иконку (close friends и т.п.).
	Privacy   string
	CreatedAt time.Time
	ExpiresAt time.Time
	// Pinned/Edited — состояние жизненного цикла истории (закреп в профиле / правка).
	Pinned, Edited bool
	// ReactionsCount — всего реакций на историю; MyReaction — эмодзи текущего
	// зрителя ("" = не реагировал); Reactions — разбивка эмодзи→count.
	ReactionsCount int
	MyReaction     string
	Reactions      []ReactionCount
	// AllowIDs — явный allow-лист истории с privacy=="selected". Заполняется
	// только для собственных историй зрителя (автор==зритель), чтобы автор мог
	// отредактировать аудиторию; для чужих историй остаётся nil (не раскрываем).
	AllowIDs []int64
	// MediaAreas — интерактивные области поверх истории (`Vector<MediaArea>`).
	MediaAreas MediaAreas
	// FwdFrom — ссылка на исходную историю при репосте (tweb fwd_from); nil у оригинала.
	FwdFrom *StoryFwd
	// Media — ВЛОЖЕНИЕ истории ступенью (`storyItem.media`). В строке `stories`
	// его нет: там лежит только номер файла, а метаданные — в `media`. Поле
	// вычисляемое, наполняет его read-модель (usecase/story) тем же приёмом,
	// каким `DialogRecord` получает последнее сообщение.
	Media MessageMedia
}

// StoryViewers — просмотры истории вместе с карточками зрителей: ровно то, из
// чего собирается контейнер `stories.storyViewsList`.
//
// Прежде наружу ехали ГОЛЫЕ карточки, и вместе с ними терялись оба параметра
// `storyView`: когда посмотрел (предмет есть — `story_views.viewed_at`) и чем
// отреагировал.
type StoryViewers struct {
	Views []StoryView
	Users []UserReal
}

// StoryGroup bundles an author's active stories for the feed read model.
type StoryGroup struct {
	Author  UserReal
	Stories []StoryRecord
	// MaxReadID — ГОРИЗОНТ прочтения зрителя у этого автора
	// (`peerStories.max_read_id`). Один номер на автора вместо признака на
	// каждой истории: «непрочитанная» это `story.id > MaxReadID`, и выводит это
	// клиент — ровно как непрочитанность сообщения по `read_inbox_max_id`.
	MaxReadID int64
}

// StealthMode — эфемерное состояние «невидимого просмотра» историй пользователя
// (tweb stories.stealthMode: active_until_date / cooldown_until_date). Пока
// now < ActiveUntil, просмотры зрителя не записываются; повторно активировать
// режим можно только после CooldownUntil.
type StealthMode struct {
	ActiveUntil   time.Time
	CooldownUntil time.Time
}

File: web-client/src/core/stories/story.ts
// Истории в форме оригинала — конструкторы схемы TL, и операции над ними.
//
// Разбор и решения Р1–Р12 — docs/readiness/tl-stories-analysis.md. До этого
// шага клиент держал свою плоскую запись (`mapStory`), в которой:
//
//  • медиа было НОМЕРОМ (`mediaId`), поэтому вьювер спрашивал mime, размеры и
//    длительность ОТДЕЛЬНЫМ запросом на каждую историю;
//  • вид аудитории и вид области были строками (`privacy`, `mediaArea.type`);
//  • прочитанность была признаком на КАЖДОЙ истории, а не горизонтом группы;
//  • реакции лежали тремя ключами верхнего уровня, из которых один
//    (`myReaction`) пер-зрительский, а два общие.
//
// Здесь ничего этого нет: форма провода и форма модели совпали, поэтому маппера
// у историй больше нет вовсе — есть операции над конструктором, как у черновика
// (`core/dialogs/draft.ts`) и у реакций сообщения
// (`core/reactions/messageReactions.ts`).
import type { MessageEntity, Reaction, ReactionCount } from '../models'
import type { GeoPoint, MessageMedia } from '../media/messageMedia'
import { getMediaFromMessage } from '../media/messageMedia'
import { getPeerId, type Peer } from '../peers/peerId'
import { isChosen } from '../reactions/messageReactions'

// ── Конструкторы ────────────────────────────────────────────────────────────

/** mediaAreaCoordinates#cfc9e002 — положение области в процентах бокса медиа. */
export interface MediaAreaCoordinates {
  _: 'mediaAreaCoordinates'
  x: number
  y: number
  w: number
  h: number
  rotation: number
}

/** mediaAreaGeoPoint#cad5452d — точка на карте поверх истории. */
export interface MediaAreaGeoPoint {
  _: 'mediaAreaGeoPoint'
  coordinates: MediaAreaCoordinates
  geo: GeoPoint
}

/** mediaAreaVenue#be82db9c — место с названием. */
export interface MediaAreaVenue {
  _: 'mediaAreaVenue'
  coordinates: MediaAreaCoordinates
  geo: GeoPoint
  title: string
  address: string
}

/** mediaAreaSuggestedReaction#14455871 — наклейка «отреагируй этим». */
export interface MediaAreaSuggestedReaction {
  _: 'mediaAreaSuggestedReaction'
  pFlags?: Partial<{ dark: true; flipped: true }>
  coordinates: MediaAreaCoordinates
  reaction: Reaction
}

/** mediaAreaUrl#37381085 — кликабельная ссылка поверх истории. */
export interface MediaAreaUrl {
  _: 'mediaAreaUrl'
  coordinates: MediaAreaCoordinates
  url: string
}

/** Интерактивная область поверх истории — ОБЪЕДИНЕНИЕ, а не одна структура с
 *  полем `type` и восемью необязательными половинами. */
export type MediaArea = MediaAreaGeoPoint | MediaAreaVenue | MediaAreaSuggestedReaction | MediaAreaUrl

/** storyFwdHeader#b826e150 — атрибуция репоста: автор ССЫЛКОЙ. */
export interface StoryFwdHeader {
  _: 'storyFwdHeader'
  from?: Peer
  story_id?: number
}

/** storyViews#8d595cd6 — ОБЩИЙ (не пер-зрительский) агрегат истории. */
export interface StoryViews {
  _: 'storyViews'
  views_count: number
  reactions?: ReactionCount[]
  reactions_count?: number
}

/** Правило аудитории (`PrivacyRule`); производим четыре конструктора. */
export type StoryPrivacyRule =
  | { _: 'privacyValueAllowAll' }
  | { _: 'privacyValueAllowContacts' }
  | { _: 'privacyValueAllowCloseFriends' }
  | { _: 'privacyValueAllowUsers'; users: number[] }

/** storyItem#16a4b93c — история. */
export interface StoryItemReal {
  _: 'storyItem'
  pFlags?: Partial<{
    pinned: true
    public: true
    close_friends: true
    contacts: true
    selected_contacts: true
    edited: true
  }>
  /** Номер ВНУТРИ автора: им история и адресуется, и по нему считается
   *  горизонт прочтения. Глобального ключа наружу не выходит. */
  id: number
  date: number
  fwd_from?: StoryFwdHeader
  expire_date: number
  caption?: string
  entities?: MessageEntity[]
  /** Ступень вложения — та же, что у сообщения. Номера файла рядом больше нет. */
  media: MessageMedia
  media_areas?: MediaArea[]
  /** Аудитория; едет ТОЛЬКО автору истории. */
  privacy?: StoryPrivacyRule[]
  views?: StoryViews
  /** Реакция ЗРИТЕЛЯ — единственная пер-зрительская часть агрегата. */
  sent_reaction?: Reaction
}

/** storyItemDeleted#51e6ee4f — «историю удалили». */
export interface StoryItemDeleted {
  _: 'storyItemDeleted'
  id: number
}

export type StoryItem = StoryItemReal | StoryItemDeleted

/** Приватность истории в терминах интерфейса. Значение ВЫВОДИТСЯ из флагов
 *  конструктора: строки `privacy` на проводе больше нет. */
export type StoryPrivacy = 'everyone' | 'contacts' | 'close' | 'selected'

// ── Операции ────────────────────────────────────────────────────────────────

/** История, а не «её удалили»: ветвление по конструктору вместо `if (story)`. */
export function realStory(s: StoryItem | undefined): StoryItemReal | undefined {
  return s?._ === 'storyItem' ? s : undefined
}

/** Подпись; «подписи нет» — отсутствие параметра, а не пустая строка. */
export function storyCaption(s: StoryItem | undefined): string {
  return realStory(s)?.caption ?? ''
}

/** Секунды эпохи — те же единицы, что у сообщения. */
export function storyDate(s: StoryItem | undefined): number {
  return realStory(s)?.date ?? 0
}

export function storyExpireDate(s: StoryItem | undefined): number {
  return realStory(s)?.expire_date ?? 0
}

/**
 * Прочитана ли история — СРАВНЕНИЕ с горизонтом, а не признак на самой истории.
 *
 * У оригинала прочитанность историй выражена одним номером на автора
 * (`peerStories.max_read_id`), ровно как непрочитанность сообщения выводится из
 * `read_inbox_max_id`. Признак на каждой истории был временной формой и ушёл
 * вместе с пер-авторской нумерацией (миграция 0126).
 */
export function isStoryRead(s: StoryItem | undefined, maxReadId: number): boolean {
  const story = realStory(s)
  return story != null && story.id <= maxReadId
}

export function isStoryPinned(s: StoryItem | undefined): boolean {
  return realStory(s)?.pFlags?.pinned === true
}

export function isStoryEdited(s: StoryItem | undefined): boolean {
  return realStory(s)?.pFlags?.edited === true
}

/**
 * Вид аудитории — по флагам конструктора.
 *
 * `contacts` последним не случайно: это ДЕФОЛТ публикации, и он же ответ, когда
 * ни один флаг не выставлен (истории прошлой схемы, кадр без флагов).
 */
export function storyPrivacy(s: StoryItem | undefined): StoryPrivacy {
  const f = realStory(s)?.pFlags
  if (f?.public) return 'everyone'
  if (f?.close_friends) return 'close'
  if (f?.selected_contacts) return 'selected'
  return 'contacts'
}

/** Явный allow-лист selected-истории. Приезжает ТОЛЬКО автору — у чужой истории
 *  вектора `privacy` нет вовсе, и ответ пуст. */
export function storyAllowIds(s: StoryItem | undefined): number[] {
  const rule = realStory(s)?.privacy?.find((r) => r._ === 'privacyValueAllowUsers')
  return rule?._ === 'privacyValueAllowUsers' ? rule.users : []
}

/** Моя реакция — эмодзи или `null`. Живёт ОТДЕЛЬНЫМ параметром истории, а не
 *  внутри агрегата: тело истории одно на всех получателей. */
export function storyMyReaction(s: StoryItem | undefined): string | null {
  const r = realStory(s)?.sent_reaction
  return r?._ === 'reactionEmoji' ? r.emoticon : null
}

/** Разбивка реакций — чипы `reactionCount` общего агрегата. */
export function storyReactions(s: StoryItem | undefined): ReactionCount[] {
  return realStory(s)?.views?.reactions ?? []
}

/** Сколько реакций всего. */
export function storyReactionsCount(s: StoryItem | undefined): number {
  return realStory(s)?.views?.reactions_count ?? 0
}

/** Интерактивные области поверх истории. */
export function storyMediaAreas(s: StoryItem | undefined): MediaArea[] {
  return realStory(s)?.media_areas ?? []
}

/** Ключ автора исходной истории у репоста; `undefined` у оригинала. */
export function storyFwdAuthorId(s: StoryItem | undefined): PeerId | undefined {
  const from = realStory(s)?.fwd_from?.from
  return from ? getPeerId(from) : undefined
}

export function storyFwdStoryId(s: StoryItem | undefined): number | undefined {
  return realStory(s)?.fwd_from?.story_id
}

/**
 * Номер файла истории — ИЗ СТУПЕНИ (`photo.id` / `document.id`), а не из
 * отдельного ключа рядом. Тот же доступ, что у вложения сообщения: `storyItem`
 * подходит `getMediaFromMessage` структурно, потому что несёт `media` тем же
 * параметром.
 */
export function storyMediaId(s: StoryItem | undefined): number {
  return getMediaFromMessage(realStory(s))?.id ?? 0
}

/** Свой ли чип реакции (`chosen_order`, а не булево «моя»). */
export { isChosen }
